pkg/backup/backup_compaction_test.go

```diff
@@ -0,0 +1,269 @@
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/util"
+	"github.com/stretchr/testify/require"
+)
+
+func TestBackupCompactionTriggered(t *testing.T) {
+	defer testutils.HookGlobal(&compactForScheduledOnly, false)()
+	ctx := context.Background()
+
+	tempDir, tempDirCleanup := testutils.TempDir(t)
+	defer tempDirCleanup()
+
+	compactionThreshold := 3
+	settings := cluster.MakeTestingClusterSettings()
+	backupCompactionThreshold.Override(ctx, &settings.SV, int64(compactionThreshold))
+	_, db, cleanup := backupRestoreTestSetupEmpty(
+		t,
+		singleNode,
+		tempDir,
+		InitManualReplication,
+		base.TestClusterArgs{
+			ServerArgs: base.TestServerArgs{
+				Settings: settings,
+			},
+		},
+	)
+	defer cleanup()
+	db.Exec(t, "BACKUP INTO 'nodelocal://1/backup'")
+	var backupPath string
+	db.QueryRow(t, "SHOW BACKUPS IN 'nodelocal://1/backup'").Scan(&backupPath)
+	countBackups := func() (count int) {
+		db.QueryRow(
+			t,
+			"SELECT COUNT(DISTINCT (start_time, end_time)) FROM "+
+				"[SHOW BACKUP FROM $1 IN 'nodelocal://1/backup']",
+			backupPath,
+		).Scan(&count)
+		return
+	}
+	for i := 2; i <= compactionThreshold; i++ {
+		db.Exec(t, "BACKUP INTO LATEST IN 'nodelocal://1/backup'")
+		if i != compactionThreshold {
+			require.Equal(t, i, countBackups())
+		}
+	}
+	require.Equal(t, compactionThreshold+1, countBackups())
+}
+
+func TestBackupCompaction(t *testing.T) {
+	defer testutils.HookGlobal(&compactForScheduledOnly, false)()
+	ctx := context.Background()
+
+	compactionThreshold := 3
+	settings := cluster.MakeTestingClusterSettings()
+	backupCompactionThreshold.Override(ctx, &settings.SV, int64(compactionThreshold))
+
+	testcases := []struct {
+		name           string
+		numNodes       int
+		clusterArgs    base.TestClusterArgs
+		collectionURIs []string
+	}{
+		{
+			"single-node",
+			singleNode,
+			base.TestClusterArgs{},
+			[]string{"nodelocal://1/backup"},
+		},
+		{
+			"locality-aware-backups",
+			multiNode,
+			base.TestClusterArgs{
+				ServerArgsPerNode: map[int]base.TestServerArgs{
+					0: {
+						Locality: roachpb.Locality{Tiers: []roachpb.Tier{
+							{Key: "region", Value: "west"},
+							{Key: "az", Value: "az1"},
+							{Key: "dc", Value: "dc1"},
+						}},
+					},
+					1: {
+						Locality: roachpb.Locality{Tiers: []roachpb.Tier{
+							{Key: "region", Value: "east"},
+							{Key: "az", Value: "az1"},
+							{Key: "dc", Value: "dc2"},
+						}},
+					},
+					2: {
+						Locality: roachpb.Locality{Tiers: []roachpb.Tier{
+							{Key: "region", Value: "east"},
+							{Key: "az", Value: "az2"},
+							{Key: "dc", Value: "dc3"},
+						}},
+					},
+				},
+			},
+			[]string{
+				fmt.Sprintf(
+					"nodelocal://1/backup?COCKROACH_LOCALITY=%s",
```

Comment:
i dont think the locality aware backups case needs to run through every single combo you provide below. We merely want to test that uri resolution works for backup compaction. So, I don't think this high level dd framework is necessary-- rather, this would be much easier to read if the locality aware test was its own stand alone test.